kratos/utilities/parallel_utilities.h

```diff
@@ -0,0 +1,364 @@
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Riccardo Rossi
+//                   Denis Demidov
+//
+
+#if !defined(KRATOS_PARALLEL_UTILITIES_H_INCLUDED )
+#define  KRATOS_PARALLEL_UTILITIES_H_INCLUDED
+
+
+// System includes
+#include<iostream>
+#include<array>
+#include<vector>
+#include<tuple>
+#include<cmath>
+#include<limits>
+#include<omp.h>
+
+#include <future>
+#include <thread>
+
+// External includes
+
+
+// Project includes
+#include "includes/define.h"
+
+
+namespace Kratos
+{
+///@addtogroup KratosCore
+
+//***********************************************************************************
+//***********************************************************************************
+//***********************************************************************************
+template<class TContainerType, class TIteratorType=typename TContainerType::iterator>
+class BlockPartition
+{
+public:
+
+    static constexpr int max_allowed_threads = 128;
+
+    BlockPartition(TIteratorType it_begin,
+                   TIteratorType it_end,
+                   int Nchunks = omp_get_max_threads())
+        : mit_begin(it_begin), mit_end(it_end), mNchunks(Nchunks)
+    {
+        int NumTerms = it_end-it_begin;
+        int mBlockPartitionize = NumTerms / mNchunks;
+        mBlockPartition[0] = 0;
+        mBlockPartition[mNchunks] = NumTerms;
+        for(int i = 1; i < mNchunks; i++)
+            mBlockPartition[i] = mBlockPartition[i-1] + mBlockPartitionize ;
+
+    }
+
+    BlockPartition(TContainerType& rData,
+                   int Nchunks = omp_get_max_threads())
+        : BlockPartition(rData.begin(), rData.end(), Nchunks)
+    {}
+
+    virtual ~BlockPartition() {}
+
+    //version enabling reduction
+    //simple version
+    template <class TUnaryFunction>
+    inline void for_each(TUnaryFunction&& f)
+    {
+        #pragma omp parallel for
+        for(int i=0; i<mNchunks; ++i)
+        {
+            for (int k = mBlockPartition[i]; k < mBlockPartition[i+1]; ++k)
+            {
+                f(*(mit_begin + k)); //note that we pass a reference to the value, not the iterator
+            }
+        }
+    }
+
+    //version with reduction
+    template <class TReducer, class TUnaryFunction>
+    inline TReducer for_reduce(TUnaryFunction &&f)
+    {
+        TReducer global_reducer;
+        #pragma omp parallel for
+        for(int i=0; i<mNchunks; ++i)
+        {
+            TReducer local_reducer;
+            for (auto k = mBlockPartition[i]; k < mBlockPartition[i+1]; ++k)
+            {
+                local_reducer.LocalMerge(f(*(mit_begin + k)));
+            }
+            global_reducer.ThreadSafeMerge(local_reducer);
+        }
+        return global_reducer;
+    }
+
+private:
+
+    TIteratorType mit_begin, mit_end;
+    int mNchunks;
+    std::array<int, max_allowed_threads> mBlockPartition;
```

Comment:
If you are changing mBlockPartition to a 64bit type, I would just replace it with an array of pointers. Then you don't need `mit_begin` and `mit_end` either:
```suggestion
    int mNchunks;
    std::array<TIteratorType, max_allowed_threads> mBlockPartition;
```

And in the constructor you would do:
```cpp
        int NumTerms = it_end-it_begin;
        int mBlockPartitionize = NumTerms / mNchunks;
        mBlockPartition[0] = it_begin;
        mBlockPartition[mNchunks] = it_end;
        for(int i = 1; i < mNchunks; i++)
            mBlockPartition[i] = mBlockPartition[i-1] + mBlockPartitionize ;
```

Then inside the for_loops:
```cpp
        #pragma omp parallel for
        for(int i=0; i<mNchunks; ++i)
        {
            for (auto p = mBlockPartition[i]; p != mBlockPartition[i+1]; ++p)
            {
                f(*p);
            }
        }
```